File: CTCI/ch1/java/StringCompression.java
/*
 * Implement a mtheod to perform baisc string compression using the counts of repeated characters. 
 * Ex: aabcccccaaa -> a2b1c5a3
 * If "compressed" string doesn't become smaller than the original string, return the original string.
 */

import java.util.Arrays;

public class StringCompression {

	public static String stringCompression(String s) {

        StringBuilder compressed = new StringBuilder();
        int consecutiveLetters = 0;
        for(int i = 0; i < s.length(); i++) {

            consecutiveLetters+=1;

            // check if the next letter is the same as the current letter or if we have reached the last letter in the string
            if (i+1 >= s.length() || s.charAt(i) != s.charAt(i+1)) {
                compressed.append(s.charAt(i));
                compressed.append(consecutiveLetters);
                consecutiveLetters = 0;
            }
        }

        return compressed.length() >= s.length() ? s : compressed.toString();
	}

	public static void main(String[] args) {
        String s1 = "aabcccccaaa";
        String s2 = "aaaaaaabb";
        String s3 = "abcde";

        String out1 = stringCompression(s1);
        String out2 = stringCompression(s2);
        String out3 = stringCompression(s3);

        System.out.println(out1); // prints a2b1c5a3
        System.out.println(out2); // a7b2
        System.out.println(out3); // abcde
	}
}

/*
	Running from command line:
		1) javac *.java <--- compile java file to class file
		2) java * <--- Execute compiled Java file
*/

File: CTCI/ch1/python/stringCompression.py
'''
	Implement a mtheod to perform baisc string compression using the counts of repeated characters. 
	Ex: aabcccccaaa -> a2b1c5a3
	If "compressed" string doesn't become smaller than the original string, return the original string.
'''

def stringCompression(inputString):

	index = 0
	outString = ''
	x = 0
	while x < len(inputString):
		consecutiveCharCount = 1
		char = inputString[x]
		if x == len(inputString)-1:
			outString+=char
			break
		if char == inputString[x+1]:
			while char == inputString[x+1]:
				consecutiveCharCount+=1
				x+=1
				if x == len(inputString)-1:
					break
			outString+=char
			outString+=str(consecutiveCharCount)
			x+=1
		else:
			outString+=char
			x+=1
	if len(outString) > len(inputString):
		return inputString
	return outString


string1 = 'aabcccccaaa'
string2 = 'aaaaaaabb'
string3 = 'abcde'
print(stringCompression(string1))
print(stringCompression(string2))
print(stringCompression(string3))


'''
	prints:
		> a2b1c5a3
		> a7b2
		> abcde
'''